pkg/receive/handler.go

```diff
@@ -747,34 +755,55 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, wreqs map[e
 	// This is needed if context is canceled or if we reached success of fail quorum faster.
 	defer func() {
 		go func() {
-			for err := range ec {
+			for errReplicaFn := range ec {
+				_, err := errReplicaFn()
 				if err != nil {
 					level.Debug(tLogger).Log("msg", "request failed, but not needed to achieve quorum", "err", err)
 				}
 			}
 		}()
 	}()
 
-	var success int
+	var (
+		// success counts how many requests per replica have succedeed.
+		success = make([]int, h.options.ReplicationFactor)
+		// replicaMultiError contains all requests errors per replica.
+		replicaMultiError = make([]errutil.MultiError, h.options.ReplicationFactor)
+		// replicaGroupSuccess counts how many replica groups have succeeded.
+		// This needs to amount to the success threshold (quorum), in order to
+		// claim replication success.
+		replicaGroupSuccess int
+	)
+
 	for {
 		select {
 		case <-fctx.Done():
 			return fctx.Err()
-		case err, more := <-ec:
+		case errReplicaFn, more := <-ec:
 			if !more {
+				for i, rme := range replicaMultiError {
+					// Only if we can determine same error for all requests within replica group,
+					// we return the cause, otherwise fallback to original original.
+					errs.Add(determineWriteErrorCause(rme.Err(), replicaGroupReqs[i]))
+				}
 				return errs.Err()
 			}
+			replica, err := errReplicaFn()
 			if err == nil {
-				success++
-				if success >= successThreshold {
-					// In case the success threshold is lower than the total
-					// number of requests, then we can finish early here. This
-					// is the case for quorum writes for example.
-					return nil
+				success[replica]++
+				for i := range success {
+					if success[i] >= replicaGroupReqs[i] {
+						// If enough replica groups succeed, we can finish early (quorum
+						// is guaranteed).
+						replicaGroupSuccess++
```

Comment:
So each `replicaGroup` signifies all the requests made to replicate a batch of series to a `replica-endpoint`? And if all the requests within that `replicaGroup` succeed, we count that as a success towards ensuring quorum?

Reply:
Yes, that's the gist :+1:, there is now also second way to verify quorum, I added more details in description.